cep-????.md

```diff
@@ -0,0 +1,129 @@
+# CEP ???? - Virtual packages
+
+<table>
+<tr><td> Title </td><td> Virtual packages </td>
+<tr><td> Status </td><td> Draft </td></tr>
+<tr><td> Author(s) </td><td> Jaime Rodríguez-Guerra &lt;jaime.rogue@gmail.com&gt;</td></tr>
+<tr><td> Created </td><td> Dec 17, 2024</td></tr>
+<tr><td> Updated </td><td> Dec 17, 2024</td></tr>
+<tr><td> Discussion </td><td> https://github.com/conda/ceps/pull/103 </td></tr>
+<tr><td> Implementation </td><td> https://github.com/conda/conda/tree/24.11.1/conda/plugins/virtual_packages, https://github.com/mamba-org/mamba/blob/libmamba-2.0.5/libmamba/src/core/virtual_packages.cpp, https://github.com/conda/rattler/tree/rattler-v0.28.8/crates/rattler_virtual_packages/src </td></tr>
+</table>
+
+> The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT",
+  "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as
+  described in [RFC2119][RFC2119] when, and only when, they appear in all capitals, as shown here.
+
+## Abstract
+
+This CEP standardizes which virtual packages MUST be offered by conda solvers.
+
+## Specification
+
+A virtual package is defined as a package record with three fields: name, version and build string.
+The name MUST start with double underscore. The version and build string MUST follow the same semantics as in regular package records.
+
+### List of virtual packages
+
+In alphabetical order, every conda client MUST support the following virtual packages:
+
+- `__archspec`
+- `__cuda`
+- `__glibc`
+- `__linux`
+- `__osx`
+- `__unix`
+- `__win`
+
+#### `__archspec`
+
+This virtual package MUST be always present, with the version set to `1`. The build string SHOULD reflect the detected CPU microarchitecture. If it cannot be detected, the build string SHOULD be `0`.
+
+The build string MUST be overridable with the `CONDA_OVERRIDE_ARCHSPEC` environment variable, if set to a non-empty value.
+
+#### `__cuda`
+
+This virtual package MUST be present when the system exhibits GPU drivers compatible with the CUDA runtimes. When available, the version value MUST be set to the oldest CUDA version supported by the detected drivers (i.e. the formatted value of `libcuda.cuDriverGetVersion()`), constrained to the first two components (major and minor) and formatted as `{major}.{minor}`. The build string MUST be `0`.
+
+The version MUST be overridable with the `CONDA_OVERRIDE_CUDA` environment variable, if set to a non-empty value.
+
+#### `__glibc`
+
+This virtual package MUST be present when the native platform is `linux-*`. Its version value MUST be set to the system `libc` version, constrained to the first two components (major and minor) formatted as `{major}.{minor}`. The build string MUST be `0`.
+
+The `libc` version can be computed via:
+
+- Python's `os.confstr("CS_GNU_LIBC_VERSION")`
+- `getconf GNU_LIBC_VERSION`
+- `ldd --version` (in GLIBC distros)
+- System's `libc.so` (in MUSL distros, location not standardized)
```

Comment:
Think MUSL should be its own thing. It is versioned differently. Plus users presumably want to distinguish GLIBC built packages from MUSL ones

Reply:
With the `stdlib` infrastructure, we have the technical requirements in place to do something like:
```yaml
c_stdlib:               # [linux]
  - sysroot             # [linux]
  - musl                # [linux]
c_stdlib_version:       # [linux]
  - 2.17                # [linux]
  - 1.2                 # [linux]
zip_keys:               # [linux]
  - - c_stdlib          # [linux]
    - c_stdlib_version  # [linux]
```
The question whether we want to roll out a "conda-forge for musl" is of course a separate question.

For the purpose of this CEP, definitely `__glibc` needs to stay musl-free IMO. For future-proofing, we could add `__musl`, but no strong feelings.

Reply:
This is the GLIBC version in MUSL distros, not the MUSL version itself. Happy to adjust the wording though. See [this block in the constructor SH templates](https://github.com/conda/constructor/blob/f6f8d62125665ceb72a6f0c3280a09be1fff8534/constructor/header.sh#L36-L60) for some more context.

Reply:
Having glibc in musl is certainly not the [default](https://wiki.alpinelinux.org/wiki/Running_glibc_programs), I actually didn't know this is supported - TIL. In that case I agree it makes sense to report the glibc that's found.

So my point was that we could also build musl-native libs, but I guess that's much less urgent if alpine-users can just install a glibc compat (assuming this works with our glibc-flavoured packages).

Reply:
Reading the original comment ( https://github.com/conda/constructor/issues/850#issuecomment-2343756454 ), think that user is referring to [`gcompat`]( https://github.com/Stantheman/gcompat?tab=readme-ov-file#introduction ):

> ...a library which provides glibc-compatible APIs for use on musl libc systems.

While that can be useful, think we should not encode that in this CEP

It may make sense to have a CEP extending this for MUSL and this could be one detail we include there, but would suggest we leave it out of this version and save it as future work

Reply:
I think I need to reword these suggestions. They are not meant to tell anyone where to get their versions from, merely inform that's where some of the existing implementations have found them, as a suggestion...

Reply:
> I think that user is referring to [gcompat](https://github.com/Stantheman/gcompat?tab=readme-ov-file&rgh-link-date=2024-12-17T23%3A01%3A03Z#introduction):

I'm not sure the original report was about `gcompat`, but about this alpine-focused [`glibc` version](https://github.com/sgerrand/alpine-pkg-glibc).

Reply:
The challenge is if we include this line (even as an example) future readers will look at this line and say "Hey that is how they said we should be supporting MUSL." Even if that is not what we meant. So taking this line out is safer than keep it IMO

The rest of the examples above seem fine

Reply:
> @jakirkham: think that user is referring to [`gcompat`](https://github.com/Stantheman/gcompat?tab=readme-ov-file#introduction):

Look at the alpine wiki link I posted:
> [wiki] If you want to run [glibc](https://www.gnu.org/software/libc/) programs in Alpine Linux, there are a few ways of doing so.

gcompat is one of them, but there are others.

> @jakirkham: future readers will look at this line and say "Hey that is how they said we should be supporting MUSL."

I don't think that's realistic; it's also easy to preempt - just say "the version of glibc, if present (musl-based distros like alpine only provide this for compatibility; their main C library is musl itself)"

Reply:
Sorry Axel think we had posted around the same time. Agree there can be other ways people try to handle this need

<hr>

To loop back to Jaime's original point on how to word this, what if we just say the following

```suggestion
- System's GNU `libc.so` (see specific disto for location)
```

Note: Some distros put `libc.so` in different places. So this is worth stating in its own right

Reply:
I reworded these paragraphs a bit including the intent of the suggestions, let me know what you think!